### PR TITLE
fix: change behaviour of maxCount in ExposedQueue

### PR DESCRIPTION
**Pull request by joachimdekker**
## Description

This PR proposes to change the behavour of maxcount in the exposed queue constructor. 

Fixes AP-172.

## Testability

Run the tests :)

## Checklist
- [x] Set the proper pull request name
- [x] Merged main into your branch

### CODE REVIEW

File: Aplib.Core/Collections/ExposedQueue.cs [hunk 1]
```diff
@@ -45,30 +46,27 @@ public ExposedQueue(int size)
...
-            if (count < 0)
-                throw new ArgumentOutOfRangeException(nameof(count), "Count cannot be negative.");
+            if (maxCount < 0) throw new ArgumentOutOfRangeException(nameof(maxCount), "Count cannot be negative.");
 
-            MaxCount = array.Length;
-            _array = array;
-            _head = MaxCount - 1;
-            Count = count;
+            MaxCount = maxCount;
+            _array = array.Take(maxCount).ToArray();
```
Comment (JensSteenmetz) [inline Aplib.Core/Collections/ExposedQueue.cs]: `array.Take(maxCount).ToArray()` creates an array of `Min(maxCount, array.Length)` elements. However, we would like to have `maxCount` elements in the resulting array, where only `array.Length` slots/cells are filled.

**Review by JensSteenmetz — CHANGES_REQUESTED** [review]

File: Aplib.Core/Collections/ExposedQueue.cs (hunk 1 above)
Comment (joachimdekker) [reply Aplib.Core/Collections/ExposedQueue.cs]: Excellent!

File: Aplib.Core/Collections/ExposedQueue.cs [hunk 2]
```diff
@@ -45,30 +46,31 @@ public ExposedQueue(int size)
...
-            _array = array;
-            _head = MaxCount - 1;
-            Count = count;
+            MaxCount = maxCount;
+            _array = new T[maxCount];
+
+            if (array.Length > maxCount) array = array[..maxCount];
+
+            array.CopyTo(_array, 0);
+            Count = array.Length;
```
Comment (JensSteenmetz) [inline Aplib.Core/Collections/ExposedQueue.cs]: If array.Length > maxCount, the array is copied twice. Can this be done more efficiently?
For example 
```suggestion
            Count = Math.Min(array.Length, MaxCount);
            
            _array = new T[maxCount];
            Array.Copy(array, _array, Count);
```

**Review by JensSteenmetz — CHANGES_REQUESTED** [review]
sorry :)

**Review by JensSteenmetz — DISMISSED** [review]
LGTM!

File: Aplib.Core/Collections/ExposedQueue.cs [hunk 3]
```diff
@@ -45,30 +46,30 @@ public ExposedQueue(int size)
...
-        /// By default, assumes the array is filled. 
+        /// By default, assumes the array is filled.
         /// </summary>
         /// <param name="array">An array to use as the circular array.</param>
-        /// <param name="count">The number of actual elements in the array.</param>
+        /// <param name="maxCount">The number of actual elements in the array.</param>
         /// <remarks>
         /// The MaxCount of the queue will be set to the length of the array.
         /// If the array is not fully filled, the Count should be specified.
         /// </remarks>
```
Comment (JoenvandeVorle) [inline Aplib.Core/Collections/ExposedQueue.cs]: Documentation is out of date now, should be updated

File: Aplib.Core/Collections/ExposedQueue.cs [hunk 4]
```diff
@@ -45,30 +46,30 @@ public ExposedQueue(int size)
...
-            if (count < 0)
-                throw new ArgumentOutOfRangeException(nameof(count), "Count cannot be negative.");
+            if (maxCount < 0) throw new ArgumentOutOfRangeException(nameof(maxCount), "Count cannot be negative.");
 
-            MaxCount = array.Length;
-            _array = array;
-            _head = MaxCount - 1;
-            Count = count;
+            MaxCount = maxCount;
+            Count = Math.Min(array.Length, maxCount);
```
Comment (JoenvandeVorle) [inline Aplib.Core/Collections/ExposedQueue.cs:64]: It seems now the constructor always assumes the array is full, there is no way to specify that you are passing an array where only x slots are filled.
How do we want this to handle empty arrays? (Open question, I don't know the best way)

**Review by JoenvandeVorle — CHANGES_REQUESTED** [review]
left 2 comments :)

File: Aplib.Core/Collections/ExposedQueue.cs (hunk 4 above)
Comment (JensSteenmetz) [reply Aplib.Core/Collections/ExposedQueue.cs:64]: I'd say we don't handle them. If I'm passing an empty array, I'd expect the resulting ExposedQueue to have those empty spots as well. It doesn't break the code, so I'd leave it like this.

Comment (JoenvandeVorle) [reply Aplib.Core/Collections/ExposedQueue.cs:64]: What happens is opposite though, it assumes the array is full even if you pass an empty one (the count just gets set to the size of the array). I talked it over with Joachim though and it seems okay to expect full arrays as long as that's clear from the summary.

**Review by JoenvandeVorle — APPROVED** [review]
lgtm

**Review by JensSteenmetz — APPROVED** [review]
LGTM

**Pull request merged**